Simulate a genome with a random genotype derived from a VCF file with allele frequencies.

Right now, it only adds single-nucleotide variants.


Installation
------------

    git clone http://github.com/jwanglab/sim_genome
    cd sim_genome
    mkdir incl
    cd incl
    git clone http://github.com/attractivechaos/klib
    cd ..
    make
    ./sg -h


Usage
-----

    ./sg ref.fasta snps.vcf [random seed] > output.fasta


Examples
--------

This example should run in a couple of minutes (after data is downloaded) and use just over 3GB of RAM.

    wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz
    wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.fa.gz
    ./sg hg19.fa.gz ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz > hg19_1kgenomes_genotype.fasta

If no random seed is given, the current epoch is used and is printed to stderr.

To simulate 100x 2x100 reads from a (heterozygous) individual (whose haplotypes are generated by random seeds 1 and 2), using [wgsim](http://github.com/lh3/wgsim):

    ./sg hg19.fa.gz ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz 1 > hg19_1kg_1.fasta
    ./sg hg19.fa.gz ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz 2 > hg19_1kg_2.fasta
    cat hg19_1kg_1.fasta hg19_1kg_2.fasta > hg19_1kg_1_2.fasta
    wgsim -N 1500000000 -1 100 -2 100 -r 0.005 hg19_1kg_1_2.fasta R1.fastq R2.fastq

To simulate an ~100x exome from individual "1\_2", requires a BED file containing exome-capture regions and [seqtk](http://github.com/lh3/seqtk):

    seqtk subseq hg19_1kg_1.fasta exome.bed > hg19_1kg_1_exome.fasta
    seqtk subseq hg19_1kg_2.fasta exome.bed > hg19_1kg_2_exome.fasta
    cat hg19_1kg_1_exome.fasta hg19_1kg_2_exome.fasta > hg19_1kg_1_2_exome.fasta
    wgsim -N 35000000 -1 100 -2 100 -r 0.005 hg19_1kg_1_2_exome.fasta exome_R1.fastq exome_R2.fastq
